gmeta/codegen/src/lib.rs

```diff
@@ -1,518 +1,503 @@
-                        pat_ident.attrs.is_empty(),
-                        &pat_ident.attrs[0],
-                        "Pattern idents shouldn't have attributes"
-                    );
-                    ensure!(
-                        pat_ident.by_ref.is_none(),
-                        &pat_ident.by_ref,
-                        "Pattern idents shouldn't be thrown as ref"
-                    );
-                    ensure!(
-                        pat_ident.subpat.is_none(),
-                        &pat_ident.subpat.as_ref().expect("checked in cond").1,
-                        "Pattern idents shouldn't be thrown using subpattern"
-                    );
-
-                    if (pat_ident.ident == "state") != state_name {
-                        return Err(compile_error(
-                            &pat_ident.ident,
-                            if state_name {
-                                "Illegal name: should be `state`"
-                            } else {
-                                "Illegal name: shouldn't be `state`"
-                            },
-                        ));
-                    }
-                } else {
-                    return Err(compile_error(
-                        pat_type,
-                        "Illegal pattern type: use common `var: Type`",
-                    ));
-                }
+        let mut items = items.into_iter();
+        let two_first_items = (items.next(), items.next());
 
-                Ok(())
-            }
-            _ => Err(compile_error(arg, "Self arguments are restricted")),
+        if let (Some(first), Some(second)) = two_first_items {
+            (first, iter::once(second).chain(items))
+        } else {
+            error!(
+                module_span,
+                "module with #[metawasm] must contain at least 2 items"
+            );
         }
+    } else {
+        error!(
+            module_span,
+            "`#[metawasm]` doesn't work with modules without a body"
+        );
     };
-    let mutate_and_validate_state =
-        |arg: &mut FnArg, state: &syn::Type| -> Result<(), TokenStream> {
-            validate_typed(arg, true)?;
-
-            if let FnArg::Typed(pat_type) = arg {
-                if let syn::Type::Path(type_path) = pat_type.ty.as_ref() {
-                    // TODO: replace with normal comparison
-                    if type_path.to_token_stream().to_string() != "Self :: State" {
-                        return Err(compile_error(
-                            type_path,
-                            "Incorrect state type: should be `Self::State`",
-                        ));
-                    }
-                } else {
-                    unreachable!("Guaranteed by `validate_typed`");
-                }
 
-                *pat_type.ty.as_mut() = state.clone();
-            } else {
-                unreachable!("Guaranteed by `validate_typed`");
-            }
+    // Checking the `State` type
 
```

Comment:
```suggestion
```

Reply:
I prefer to leave the empty line after comment related to a code section. It distinguishes them from comments on particular statements/items/lines.

Reply:
ok, leave it as it is